### PR TITLE
Feature/gradient clipping

### PR DESCRIPTION
**Pull request by pldelisle**

### CODE REVIEW

**Comment by pldelisle** [comment]
Updated according to what we discussed.

File: kerosene/metrics/metrics.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ class Dice(Metric):
     SUPPORTED_REDUCTIONS = [None, "mean"]
 
     def __init__(self, num_classes: int, reduction: Union[None, str] = "mean", average: str = None,
-                 ignore_index: int = -100,
+                 weight: List[torch.Tensor] = None, ignore_index: int = -100,
```
Comment (banctilrobitaille) [inline kerosene/metrics/metrics.py]: Why a list of tensor instead of a tensor ?

File: kerosene/nn/apex.py [hunk 2]
```diff
@@ -56,39 +58,95 @@ def backward(self, gradient: Optional[Tensor] = None, retain_graph=False,
             self._loss.backward(gradient, retain_graph, create_graph)
 
     def mean(self):
-        self._loss = torch.mean(self._loss)
-        return self
+        loss = torch.mean(self._loss)
```
Comment (banctilrobitaille) [inline kerosene/nn/apex.py]: would def need tests on these operation as they are crucial. :point_down: I  see a potential GC overhead limit error (cause of potential of huge number of object to clear anmd quick) here but we'll see.

File: kerosene/nn/criterions.py [hunk 3]
```diff
@@ -218,6 +222,139 @@ def ignore_index_fn(dice_vector):
         return dice
 
 
+class TverskyLoss(_Loss):
```
Comment (banctilrobitaille) [inline kerosene/nn/criterions.py]: Are the twersky loss tested ? Where is the implementation coming from ?

File: kerosene/nn/init.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from enum import Enum
+from typing import Union, Callable
+
+from torch.nn import init
+
+
+class InitializerType(Enum):
```
Comment (banctilrobitaille) [inline kerosene/nn/init.py]: I think this is out of the scope of the library as its model related.

File: kerosene/training/trainers.py [hunk 5]
```diff
@@ -183,31 +221,39 @@ def update_valid_metric(self, metric):
...
-        self._status = Status.FINALIZE
+        self._status = Status.FINALIZING
 
 
 class Trainer(EventGenerator):
     LOGGER = logging.getLogger("Trainer")
 
-    def __init__(self, name, train_data_loader: DataLoader, valid_data_loader: DataLoader,
-                 model_trainers: Union[List[ModelTrainer], ModelTrainer],
+    def __init__(self, name, train_data_loader: Union[None, DataLoader], valid_data_loader: Union[None, DataLoader],
```
Comment (banctilrobitaille) [inline kerosene/training/trainers.py]: How is it possible not to have Dataloaders ?

File: kerosene/training/trainers.py [hunk 6]
```diff
@@ -383,6 +515,13 @@ def validate_step(self, inputs, target):
         model.compute_valid_metric(pred, target)
         model.compute_and_update_valid_loss(pred, target)
 
+    def test_step(self, inputs, target):
+        model = self._model_trainers[0]
+
+        pred = model.forward(inputs)
+        model.compute_valid_metric(pred, target)
```
Comment (banctilrobitaille) [inline kerosene/training/trainers.py]: shouldn't it be test metric and loss ?

File: kerosene/training/trainers.py [hunk 7]
```diff
@@ -286,28 +344,44 @@ def _at_least_one_model_is_active(self):
         return len(list(filter(lambda model_trainer: model_trainer.is_active(), self._model_trainers))) > 0
 
     def train(self, nb_epoch):
+        self._on_training_begin()
         self.fire(Event.ON_TRAINING_BEGIN)
 
         for self._current_epoch in range(0, nb_epoch):
             if self._at_least_one_model_is_active():
                 self._on_epoch_begin()
+                self.fire(Event.ON_EPOCH_BEGIN)
```
Comment (banctilrobitaille) [inline kerosene/training/trainers.py]: Why not let the fire in the private method ? At the end

File: kerosene/config/trainers.py [hunk 8]
```diff
@@ -145,9 +161,9 @@ def metric_params(self):
     @classmethod
     def from_dict(cls, model_name, config_dict):
         return cls(model_name, config_dict["type"], config_dict["params"], config_dict["optimizer"]["type"],
-                   config_dict["optimizer"]["params"], config_dict["scheduler"]["type"],
-                   config_dict["scheduler"]["params"], config_dict["criterion"]["type"],
-                   config_dict["criterion"]["params"], config_dict["metric"]["type"],
+                   config_dict["optimizer"]["max_grad_norm"], config_dict["optimizer"]["params"],
```
[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/config/trainers.py]: Moved one step above in dict hierarchy. Would now be: 

models:
  Segmenter:
    name: "UNet3D"
    type: "UNet3D"
    params:
      ...
    optimizer:
      type: "FusedSGD"
      params:
        ...
    scheduler:
      type: "ReduceLROnPlateau"
      params:
        ...
    criterion:
      ...
    metric:
      ...
    params:
        max_grad_norm: 1.0

[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/config/trainers.py]: Can't put it anywhere else I think. Cannot be in 
models:
   type:...
   name:...
   params:
      max_grad_norm

because these params are passed directly to class initializer of the model. 

Could be in the params of the optimizer, but again these params are directly passed when instantiating the optimizer.

File: kerosene/dataloaders/dataloaders.py [hunk 9]
```diff
@@ -9,21 +9,24 @@
 
 
 class DataloaderFactory(object):
-    def __init__(self, train_dataset: Dataset, valid_dataset: Dataset):
+    def __init__(self, train_dataset: Dataset, valid_dataset: Dataset, test_dataset: Dataset):
```
[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/dataloaders/dataloaders.py]: `torch.distributed.init_process_group(backend='nccl', init_method='env://', rank=run_config.local_rank)`
Would have to be in the user's main. Do we want the user call this and the logic of samplers when using distributed learning ? This is the real question.

Shuffle is mutually exclusive when using samplers. You cannot set it to True if using DistributedSampler.

File: kerosene/dataloaders/dataloaders.py [hunk 10]
```diff
@@ -9,21 +9,24 @@
...
         self._valid_dataset = valid_dataset
+        self._test_dataset = test_dataset
 
     def create(self, run_config: RunConfiguration, training_config: TrainerConfiguration, collate_fn: Callable = None):
         devices = run_config.devices
         if not on_single_device(devices):
             torch.distributed.init_process_group(backend='nccl', init_method='env://', rank=run_config.local_rank)
             train_sampler = torch.utils.data.distributed.DistributedSampler(self._train_dataset)
             valid_sampler = torch.utils.data.distributed.DistributedSampler(self._valid_dataset)
+            test_sampler = torch.utils.data.distributed.DistributedSampler(self._test_dataset)
```
[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/dataloaders/dataloaders.py]: With some sort of Lambda sorcery ?

File: kerosene/events/generators/base_generator.py [hunk 11]
```diff
@@ -40,17 +40,73 @@ def fire(self, event: BaseEvent):
                 handler(event, state)
 
     @abstractmethod
-    def _on_epoch_begin(self):
+    def on_epoch_begin(self):
```
[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/events/generators/base_generator.py]: Misunderstood what you were meaning. Corrected now.

File: kerosene/nn/criterions.py (hunk 3 above)
Comment (pldelisle) [reply kerosene/nn/criterions.py]: Implementation is exactly the same as the Dice Loss, but with two coefficients more (alpha and beta). 
But no, not tested. Will write a test with alpha and beta = 0.5 which corresponds to the Dice Loss. Otherwise, would be a bit hard to test.

File: kerosene/nn/init.py (hunk 4 above)
Comment (pldelisle) [reply kerosene/nn/init.py]: Models are in SAMITorch. What if the user want different initialization ? Should be initialize them in SAMITorch or let the user init the nets with Kerosene ?

File: kerosene/events/generators/base_generator.py [hunk 12]
```diff
@@ -40,17 +40,73 @@ def fire(self, event: BaseEvent):
...
 
     @abstractmethod
-    def _on_epoch_end(self):
+    def on_epoch_end(self):
         raise NotImplementedError()
 
     @abstractmethod
-    def on_epoch_begin(self):
-        pass
+    def on_validation_epoch_begin(self):
```
[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/events/generators/base_generator.py]: I personally hate diminutives and abbreviations. But I can change it ...

File: kerosene/nn/apex.py (hunk 2 above)
Comment (pldelisle) [reply kerosene/nn/apex.py]: I see. But it's the only way I can think that doesn't change the value after in user's space. It's the only way I can thing to get this correct behaviour.

File: kerosene/nn/criterions.py (hunk 3 above)
Comment (pldelisle) [reply kerosene/nn/criterions.py]: Test done.

File: kerosene/config/trainers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pldelisle) [reply kerosene/config/trainers.py]: Created a section "gradients" in config file.

File: kerosene/dataloaders/dataloaders.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (banctilrobitaille) [reply kerosene/dataloaders/dataloaders.py]: Abstract Application class, as discussed.
